analysis.go

```diff
@@ -0,0 +1,164 @@
+func NewAnalyzer(c *Config) *analysis.Analyzer {
+	var initOncer sync.Once
+	var initErr error
+	var goheader *Analyzer
+
+	return &analysis.Analyzer{
+		Doc:              "the_only_doc",
+		URL:              "https://github.com/denis-tingaikin/go-header",
+		Name:             "goheader",
+		RunDespiteErrors: true,
+		Run: func(p *analysis.Pass) (any, error) {
+			initOncer.Do(func() {
+				var templ string
+				var vals map[string]Value
+
+				templ, initErr = c.GetTemplate()
+				if initErr != nil {
+					return
+				}
+
+				vals, initErr = c.GetValues()
+				if initErr != nil {
+					return
+				}
+
+				goheader = New(WithTemplate(templ), WithValues(vals))
+
+			})
+			if initErr != nil {
+				return nil, initErr
+			}
+
+			var wg sync.WaitGroup
+
+			var jobCh = make(chan *ast.File, len(p.Files))
+
+			for _, f := range p.Files {
+				file := f
+				jobCh <- file
+			}
+			close(jobCh)
+
+			for range runtime.NumCPU() {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+
+					for file := range jobCh {
+						filename := p.Fset.Position(file.Pos()).Filename
+						if !strings.HasSuffix(filename, ".go") {
+							continue
+						}
+
+						res := goheader.Analyze(&Target{
+							File: file,
+							Path: filename,
+						})
+
+						if res.Err == nil {
+							continue
+						}
+						var line = 1
+						if hasCGOImport(file) {
+							line = 4
+						}
+
+						var start = p.Fset.File(file.FileStart).LineStart(line)
+						var end = res.End - res.Start + start
+						var endLine = p.Fset.File(file.FileStart).Line(end) + 1
+						end = p.Fset.File(file.FileStart).LineStart(endLine)
+
+						diag := analysis.Diagnostic{
+							Pos:     start,
+							End:     end,
+							Message: res.Err.Error(),
+						}
+
+						if res.Fix != "" {
+							diag.SuggestedFixes = []analysis.SuggestedFix{{
+								TextEdits: []analysis.TextEdit{{
+									Pos:     start,
+									End:     end,
+									NewText: []byte(res.Fix + "\n"),
+								}},
+							}}
+						}
+						p.Report(diag)
+					}
+				}()
+			}
+
+			wg.Wait()
+			return nil, nil
+		},
+	}
+}
+func hasCGOImport(file *ast.File) bool {
+	var unsafeCount int
+	for _, imp := range file.Imports {
+		if imp.Path.Value == `"C"` {
+			return true
+		}
+		if imp.Path.Value == `"unsafe"` {
+			unsafeCount++
+		}
+	}
+	return unsafeCount > 1
+}
+
+// NewAnalyzerFromConfigPath creates a new analysis.Analyzer from goheader config file
+func NewAnalyzerFromConfigPath(config *string) *analysis.Analyzer {
+	var goheaderOncer sync.Once
+	var goheader *analysis.Analyzer
+
+	return &analysis.Analyzer{
+		Doc:              "the_only_doc",
+		URL:              "https://github.com/denis-tingaikin/go-header",
+		Name:             "goheader",
+		RunDespiteErrors: true,
+		Run: func(p *analysis.Pass) (any, error) {
+			var err error
+			goheaderOncer.Do(func() {
+				var cfg Config
+				if err = cfg.Parse(*config); err != nil {
+					return
+				}
+				goheader = NewAnalyzer(&cfg)
+			})
```

Comment:
you can move this before the creation of the `analysis.Analyzer`

Reply:
Thansk. Will be fixed in the separate PR